In [4]:
# Synthetic dataset
from sklearn.datasets import make_classification
# Data processing
import pandas as pd
import numpy as np
from collections import Counter
# Visualization
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn
# Model and performance
from sklearn.svm import OneClassSVM
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.datasets import load_iris
from sklearn.datasets import load_breast_cancer


In [237]:
#DataSets that we can test
#climate
liver = 'https://archive.ics.uci.edu/ml/machine-learning-databases/liver-disorders/bupa.data'
arrhythmia = 'https://archive.ics.uci.edu/ml/machine-learning-databases/arrhythmia/arrhythmia.data'
vowel = 'https://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/vowel/vowel-context.data'
iris = load_iris()
breast_cancer = load_breast_cancer()

name_datasets_csv = {'arrhythmia': arrhythmia}
name_datasets_sklearn = {'iris_data': iris, 'breast_cancer_data': breast_cancer}


In [14]:
data = {}
target = {}
dimensionality = {}
datasets = {}


In [15]:
vowel_data = pd.read_csv(vowel,skiprows=2, index_col=False,delim_whitespace=True)
vowel_data.columns = ['feature{}'.format(i) for i in range(0,vowel_data.shape[1])]

dimensionality['vowel_data'] = [vowel_data.shape,vowel_data.nunique().tolist()[-1:]]
data['vowel_data'] = vowel_data[vowel_data.columns[:-1]]
target['vowel_data'] = vowel_data[vowel_data.columns[-1:]]

In [105]:
data['vowel_data']

feature0  feature1  feature2  feature3  feature4  feature5  feature6  \
0           0         0         0    -2.287     1.809    -1.498     1.012   
1           0         0         0    -2.598     1.938    -0.846     1.062   
2           0         0         0    -2.852     1.914    -0.755     0.825   
3           0         0         0    -3.482     2.524    -0.433     1.048   
4           0         0         0    -3.941     2.305     0.124     1.771   
..        ...       ...       ...       ...       ...       ...       ...   
982         1        14         1    -3.239     3.083    -1.427    -0.202   
983         1        14         1    -3.753     3.605    -0.899    -0.747   
984         1        14         1    -3.980     2.459     0.068     0.023   
985         1        14         1    -4.264     2.925     0.065     0.794   
986         1        14         1    -3.291     2.324    -0.679     0.285   

     feature7  feature8  feature9  feature10  feature11  feature12  
0      -1.053     1.060    -0.567      0.235     -0.091     -0.795  
1      -1.633     0.764     0.394     -0.150      0.277     -0.396  
2      -1.588     0.855     0.217     -0.246      0.238     -0.365  
3      -1.995     0.902     0.322      0.450      0.377     -0.366  
4      -1.815     0.593    -0.435      0.992      0.575     -0.301  
..        ...       ...       ...        ...        ...        ...  
982    -0.282     1.421     0.576      0.068     -0.914      0.147  
983    -0.401     1.765     0.620      0.754     -0.835     -0.301  
984     0.237     1.029    -0.189      0.521     -0.773     -0.500  
985     0.323     0.515    -1.282     -0.140     -0.863     -0.390  
986     0.441     0.557    -0.227      0.115     -1.046      0.697  

[987 rows x 13 columns]

In [26]:
liver_data = pd.read_csv(liver,skiprows=2, index_col=False)
liver_data.columns = ['feature{}'.format(i) for i in range(0,liver_data.shape[1])]
liver_data = liver_data[liver_data.columns[:-1]]

dimensionality['liver'] = [liver_data.shape,liver_data.nunique().tolist()[-1:]]
data['liver'] = liver_data[liver_data.columns[:-1]]
target['liver'] = liver_data[liver_data.columns[-1:]]
datasets['liver'] = liver_data

In [30]:
liver_data

feature0  feature1  feature2  feature3  feature4  feature5
0          91        78        34        24        36       0.0
1          87        70        12        28        10       0.0
2          98        55        13        17        17       0.0
3          88        62        20        17         9       0.5
4          88        67        21        11        11       0.5
..        ...       ...       ...       ...       ...       ...
337        99        75        26        24        41      12.0
338        96        69        53        43       203      12.0
339        98        77        55        35        89      15.0
340        91        68        27        26        14      16.0
341        98        99        57        45        65      20.0

[342 rows x 6 columns]

In [238]:
#csv
for name in name_datasets_csv:
    var = name+"_data"
    var = pd.read_csv(str(name_datasets_csv.get(name)),skiprows=2, index_col=False)
    var.columns = ['feature{}'.format(i) for i in range(0,var.shape[1])]

    dimensionality[name] = [var.shape,var.nunique().tolist()[-1:]]
    data[name] = var[var.columns[:-1]]
    target[name] = var[var.columns[-1:]]
    datasets[name] = var
    


In [239]:
#arrhythmia
data['arrhythmia']

feature0  feature1  feature2  feature3  feature4  feature5  feature6  \
0          55         0       175        94       100       202       380   
1          75         0       190        80        88       181       360   
2          13         0       169        51       100       167       321   
3          40         1       160        52        77       129       377   
4          49         1       162        54        78         0       376   
..        ...       ...       ...       ...       ...       ...       ...   
444        53         1       160        70        80       199       382   
445        37         0       190        85       100       137       361   
446        36         0       166        68       108       176       365   
447        32         1       155        55        93       106       386   
448        78         1       160        70        79       127       364   

     feature7  feature8  feature9  ... feature269 feature270 feature271  \
0         179       143        28  ...        0.1        0.0       12.2   
1         177       103       -16  ...       -0.4        0.0       13.1   
2         174        91       107  ...        0.0       -0.6       12.2   
3         133        77        77  ...       -0.4        0.0        6.5   
4         157        70        67  ...       -0.3        0.0        8.2   
..        ...       ...       ...  ...        ...        ...        ...   
444       154       117       -37  ...        0.0        0.0        4.3   
445       201        73        86  ...       -0.5        0.0       15.6   
446       194       116       -85  ...        1.2        0.0       16.3   
447       218        63        54  ...        0.2       -0.4       12.0   
448       138        78        28  ...       -0.3        0.0       10.4   

    feature272 feature273  feature274  feature275  feature276  feature277  \
0         -2.2        0.0         0.0         0.4         2.6        34.6   
1         -3.6        0.0         0.0        -0.1         3.9        25.4   
2         -2.8        0.0         0.0         0.9         2.2        13.5   
3          0.0        0.0         0.0         0.4         1.0        14.3   
4         -1.9        0.0         0.0         0.1         0.5        15.8   
..         ...        ...         ...         ...         ...         ...   
444       -5.0        0.0         0.0         0.7         0.6        -4.4   
445       -1.6        0.0         0.0         0.4         2.4        38.0   
446      -28.6        0.0         0.0         1.5         1.0       -44.2   
447       -0.7        0.0         0.0         0.5         2.4        25.0   
448       -1.8        0.0         0.0         0.5         1.6        21.3   

     feature278  
0          61.6  
1          62.8  
2          31.1  
3          20.5  
4          19.8  
..          ...  
444        -0.5  
445        62.4  
446       -33.2  
447        46.6  
448        32.8  

[449 rows x 279 columns]

In [37]:
#sklearn
for name in name_datasets_sklearn:

    var = name+"_data"
    df = name_datasets_sklearn.get(name)

    var = pd.DataFrame(data=data, columns=name_datasets_sklearn.get(name).feature_names)
    var.columns = ['feature{}'.format(i) for i in range(0,var.shape[1])]
    
    data[name] = pd.DataFrame(df.data)
    target[name] = pd.DataFrame(df.target)
    datasets[name] = pd.DataFrame(df.data).assign(target=df.target)
    dimensionality[name] = [datasets[name].shape,pd.DataFrame(df.target).nunique().tolist()[-1:]]

In [65]:
datasets['iris_data']

0    1    2    3  target
0    5.1  3.5  1.4  0.2       0
1    4.9  3.0  1.4  0.2       0
2    4.7  3.2  1.3  0.2       0
3    4.6  3.1  1.5  0.2       0
4    5.0  3.6  1.4  0.2       0
..   ...  ...  ...  ...     ...
145  6.7  3.0  5.2  2.3       2
146  6.3  2.5  5.0  1.9       2
147  6.5  3.0  5.2  2.0       2
148  6.2  3.4  5.4  2.3       2
149  5.9  3.0  5.1  1.8       2

[150 rows x 5 columns]

#### Creating Data Set for Iris data and assigning target value 2 as an outlier

In [71]:
iris_data_df = pd.DataFrame(datasets['iris_data'])

In [73]:
iris_data_df['target'] = iris_data_df['target'].apply(lambda x: 1 if x == 2 else 0 if not np.isnan(x) else 0)

In [83]:
# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(iris_data_df[iris_data_df.columns[:-1]], iris_data_df[iris_data_df.columns[-1:]], test_size=0.3, random_state=42)

In [89]:
# Print the shapes of the resulting arrays
print('X_train shape:', X_train.shape)
print('y_train shape:', y_train.shape)
print('X_test shape:', X_test.shape)
print('y_test shape:', y_test.shape)

X_train shape: (105, 4)
y_train shape: (105, 1)
X_test shape: (45, 4)
y_test shape: (45, 1)


In [90]:
# Train the one class support vector machine (SVM) model
one_class_svm = OneClassSVM(nu=0.01, kernel = 'rbf', gamma = 'auto').fit(X_train)

In [91]:
# Predict the anomalies
prediction = one_class_svm.predict(X_test)
# Change the anomalies' values to make it consistent with the true values
prediction = [1 if i==-1 else 0 for i in prediction]

In [92]:
print("The accuracy score predicted is {}".format(accuracy_score(y_test, prediction)))


The accuracy score predicted is 0.7333333333333333


In [93]:
# Check the model performance
print(classification_report(y_test, prediction))

              precision    recall  f1-score   support

           0       0.74      0.97      0.84        32
           1       0.67      0.15      0.25        13

    accuracy                           0.73        45
   macro avg       0.70      0.56      0.54        45
weighted avg       0.72      0.73      0.67        45



#### Creating Data Set for arrhythmia data and assigning target value 10 as an outlier (49 rows out of 449)

In [240]:
datasets['arrhythmia']

feature0  feature1  feature2  feature3  feature4  feature5  feature6  \
0          55         0       175        94       100       202       380   
1          75         0       190        80        88       181       360   
2          13         0       169        51       100       167       321   
3          40         1       160        52        77       129       377   
4          49         1       162        54        78         0       376   
..        ...       ...       ...       ...       ...       ...       ...   
444        53         1       160        70        80       199       382   
445        37         0       190        85       100       137       361   
446        36         0       166        68       108       176       365   
447        32         1       155        55        93       106       386   
448        78         1       160        70        79       127       364   

     feature7  feature8  feature9  ... feature270 feature271 feature272  \
0         179       143        28  ...        0.0       12.2       -2.2   
1         177       103       -16  ...        0.0       13.1       -3.6   
2         174        91       107  ...       -0.6       12.2       -2.8   
3         133        77        77  ...        0.0        6.5        0.0   
4         157        70        67  ...        0.0        8.2       -1.9   
..        ...       ...       ...  ...        ...        ...        ...   
444       154       117       -37  ...        0.0        4.3       -5.0   
445       201        73        86  ...        0.0       15.6       -1.6   
446       194       116       -85  ...        0.0       16.3      -28.6   
447       218        63        54  ...       -0.4       12.0       -0.7   
448       138        78        28  ...        0.0       10.4       -1.8   

    feature273 feature274  feature275  feature276  feature277  feature278  \
0          0.0        0.0         0.4         2.6        34.6        61.6   
1          0.0        0.0        -0.1         3.9        25.4        62.8   
2          0.0        0.0         0.9         2.2        13.5        31.1   
3          0.0        0.0         0.4         1.0        14.3        20.5   
4          0.0        0.0         0.1         0.5        15.8        19.8   
..         ...        ...         ...         ...         ...         ...   
444        0.0        0.0         0.7         0.6        -4.4        -0.5   
445        0.0        0.0         0.4         2.4        38.0        62.4   
446        0.0        0.0         1.5         1.0       -44.2       -33.2   
447        0.0        0.0         0.5         2.4        25.0        46.6   
448        0.0        0.0         0.5         1.6        21.3        32.8   

     feature279  
0             1  
1             7  
2            14  
3             1  
4             1  
..          ...  
444           1  
445          10  
446           2  
447           1  
448           1  

[449 rows x 280 columns]

In [241]:
arrhythmia_df = pd.DataFrame(datasets['arrhythmia'])

In [244]:
arrhythmia_df.rename(columns = {'feature279':'target'}, inplace = True)
arrhythmia_df

feature0  feature1  feature2  feature3  feature4  feature5  feature6  \
0          55         0       175        94       100       202       380   
1          75         0       190        80        88       181       360   
2          13         0       169        51       100       167       321   
3          40         1       160        52        77       129       377   
4          49         1       162        54        78         0       376   
..        ...       ...       ...       ...       ...       ...       ...   
444        53         1       160        70        80       199       382   
445        37         0       190        85       100       137       361   
446        36         0       166        68       108       176       365   
447        32         1       155        55        93       106       386   
448        78         1       160        70        79       127       364   

     feature7  feature8  feature9  ... feature270 feature271 feature272  \
0         179       143        28  ...        0.0       12.2       -2.2   
1         177       103       -16  ...        0.0       13.1       -3.6   
2         174        91       107  ...       -0.6       12.2       -2.8   
3         133        77        77  ...        0.0        6.5        0.0   
4         157        70        67  ...        0.0        8.2       -1.9   
..        ...       ...       ...  ...        ...        ...        ...   
444       154       117       -37  ...        0.0        4.3       -5.0   
445       201        73        86  ...        0.0       15.6       -1.6   
446       194       116       -85  ...        0.0       16.3      -28.6   
447       218        63        54  ...       -0.4       12.0       -0.7   
448       138        78        28  ...        0.0       10.4       -1.8   

    feature273 feature274  feature275  feature276  feature277  feature278  \
0          0.0        0.0         0.4         2.6        34.6        61.6   
1          0.0        0.0        -0.1         3.9        25.4        62.8   
2          0.0        0.0         0.9         2.2        13.5        31.1   
3          0.0        0.0         0.4         1.0        14.3        20.5   
4          0.0        0.0         0.1         0.5        15.8        19.8   
..         ...        ...         ...         ...         ...         ...   
444        0.0        0.0         0.7         0.6        -4.4        -0.5   
445        0.0        0.0         0.4         2.4        38.0        62.4   
446        0.0        0.0         1.5         1.0       -44.2       -33.2   
447        0.0        0.0         0.5         2.4        25.0        46.6   
448        0.0        0.0         0.5         1.6        21.3        32.8   

     target  
0         0  
1         0  
2         0  
3         0  
4         0  
..      ...  
444       0  
445       1  
446       0  
447       0  
448       0  

[449 rows x 280 columns]

In [243]:
arrhythmia_df['target'] = arrhythmia_df['target'].apply(lambda x: 1 if x == 10 else 0 if not np.isnan(x) else 0)

In [245]:
for i in range(0,278):
               arrhythmia_df['feature{}'.format(i)] = arrhythmia_df['feature{}'.format(i)].replace('?', float('nan'))

In [250]:
arrhythmia_df = arrhythmia_df.fillna(0)

In [251]:
# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(arrhythmia_df[arrhythmia_df.columns[:-1]], arrhythmia_df[arrhythmia_df.columns[-1:]], test_size=0.3, random_state=42)

In [252]:
# Print the shapes of the resulting arrays
print('X_train shape:', X_train.shape)
print('y_train shape:', y_train.shape)
print('X_test shape:', X_test.shape)
print('y_test shape:', y_test.shape)

X_train shape: (314, 279)
y_train shape: (314, 1)
X_test shape: (135, 279)
y_test shape: (135, 1)


In [253]:
# Train the one class support vector machine (SVM) model
one_class_svm = OneClassSVM(nu=0.01, kernel = 'rbf', gamma = 'auto').fit(X_train)

In [254]:
# Predict the anomalies
prediction = one_class_svm.predict(X_test)
# Change the anomalies' values to make it consistent with the true values
prediction = [1 if i==-1 else 0 for i in prediction]

In [255]:
print("The accuracy score predicted is {}".format(accuracy_score(y_test, prediction)))


The accuracy score predicted is 0.1111111111111111


In [256]:
# Check the model performance
print(classification_report(y_test, prediction))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00       120
           1       0.11      1.00      0.20        15

    accuracy                           0.11       135
   macro avg       0.06      0.50      0.10       135
weighted avg       0.01      0.11      0.02       135



C:\Users\Gokul\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Gokul\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Gokul\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
